docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,444 @@
+- Write once: Key can only be stored once and can't be removed. 
+
+Encryption, authentication & rollback protection features use an integer "unique key ID". This ID is incremented whenever we have a new key. Its initial value is determined at init time by iterating over all RBP keys in the rollback protection storage, and taking the maximal value + 1.    
+
+
+![SecureStore Layers](./SecureStore_layers.jpg)
+
+
+### Data layout
+
+When storing the data, it is stored with a preceding metadata header. Metadata includes flags and security related parameters such as nonce and unique key. The CMAC, calculated for authentication, is stored at the end of the data as it is calculated on the fly, so it can't be stored with the metadata.  
+
+![SecureStore Record](./SecureStore_record.jpg)
+
+Fields are:
+- Metadata size: Size of metadata header
+- Data size: Size of user data
+- Flags: User flags
+- Nonce: Nonce field used in IV
+- Pad: Pad data to a multiple of 16 bytes (due to encryption)
+- CMAC: CMAC calculated on key, metadata & data  
+
+### Basic implementation concepts
+
+When setting the data, as the code can't construct a single buffer to store all data (including metadata and possibly encrypted data) in one shot, it will be done in chunks, using the incremental set APIs. As for get, it will use the offset argument to extract metadata, data & CMAC separately.  
+Rollback protection (RBP) keys are stored in the designated rollback protection storage, of KVStore type also. RBP Keys are generated using a 32-bit incremental counter, which is incremented whenever we have a new key. Start value of the counter is achieved by iterating over the rollback protection KVStore (at initialization) and getting the maximal value. 
+
+ 
+# Detailed design
+
+![SecureStore Class Hierarchy](./SecureStore_class_hierarchy.jpg)
+
+Functionality, as defined by KVStore, includes the following:
+- Initialization & reset
+- Core actions: get, set & remove
+- Incremental set actions
+- Iterator actions
+
+### Class header
+
+SecureStore has the following header:
+
+```C++
+class SecureStore : KVStore {
+
+public:
+    SecureStore(KVStore *underlying_kv, KVStore *rbp_kv);
+    virtual ~SecureStore();
+    	 
+    // Initialization and formatting
+    virtual int init();
+    virtual int deinit();
+    virtual int reset();
+
+    // Core API
+    virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);
+    virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size, size_t offset = 0);
+    virtual int get_info(const char *key, info_t *info);
+    virtual int remove(const char *key);
+ 
+    // Incremental set API
+    virtual int set_start(set_handle_t *handle, const char *key, size_t final_data_size, uint32_t create_flags);
+    virtual int set_add_data(set_handle_t handle, const void *value_data, size_t data_size);
```

Comment:
Value data was decided in confluence as a term to tell it's the data of the *value* (in the Key Value store).
Regarding data size - Yes. If a SecureStore `set` function calls a TDBStore `set_add_data`, the it will be 32 (or less).